drivers/serial/uart_rzt2m.c

```diff
@@ -0,0 +1,439 @@
+{
+	const struct rzt2m_device_config *config = dev->config;
+
+	if (FTSR_T(*FTSR(config->base)) == MAX_FIFO_DEPTH ||
+	    ((*CCR0(config->base) & CCR0_MASK_TIE) == 0)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+static int uart_rzt2m_irq_rx_ready(const struct device *dev)
+{
+	const struct rzt2m_device_config *config = dev->config;
+
+	if (FRSR_R(*FRSR(config->base))) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int uart_rzt2m_irq_is_pending(const struct device *dev)
+{
+	const struct rzt2m_device_config *config = dev->config;
+
+	if ((*CSR(config->base) & (CSR_MASK_RDRF)) || (*FRSR(config->base) & FRSR_MASK_DR)) {
+		return 1;
+	}
+	return 0;
+}
+
+static void uart_rzt2m_irq_callback_set(const struct device *dev, uart_irq_callback_user_data_t cb,
+					void *cb_data)
+{
+	struct rzt2m_device_data *data = dev->data;
+
+	data->callback = cb;
+	data->callback_data = cb_data;
+}
+
+static int uart_rzt2m_irq_update(const struct device *dev)
+{
+	const struct rzt2m_device_config *config = dev->config;
+
+	*CFCLR(config->base) = CFCLR_MASK_RDRFC;
+	*FFCLR(config->base) = FFCLR_MASK_DRC;
+	return 1;
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api rzt2m_uart_api = {
+	.poll_in = rzt2m_poll_in,
+	.poll_out = rzt2m_poll_out,
+	.err_check = rzt2m_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = rzt2m_fifo_fill,
+	.fifo_read = rzt2m_fifo_read,
+	.irq_rx_enable = uart_rzt2m_irq_rx_enable,
+	.irq_rx_disable = uart_rzt2m_irq_rx_disable,
+	.irq_tx_enable = uart_rzt2m_irq_tx_enable,
+	.irq_tx_disable = uart_rzt2m_irq_tx_disable,
+	.irq_tx_ready = uart_rzt2m_irq_tx_ready,
+	.irq_rx_ready = uart_rzt2m_irq_rx_ready,
+	.irq_is_pending = uart_rzt2m_irq_is_pending,
+	.irq_callback_set = uart_rzt2m_irq_callback_set,
+	.irq_update = uart_rzt2m_irq_update,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int rzt2m_module_start(const struct device *dev)
+{
+	if (!dev || !dev->config || !dev->data) {
+		return -ENODEV;
+	}
+
+	const struct rzt2m_device_config *config = dev->config;
+	struct rzt2m_device_data *data = dev->data;
+	int interface_id = BASE_TO_IFACE_ID(config->base);
+	unsigned int irqkey = irq_lock();
+	volatile uint32_t dummy;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if (interface_id < 5) {
+		/* Dummy-read at least one time as stated in 8.3.1 of the User's Manual: Hardware */
+		*MSTPCRA &= ~(MSTPCRA_MASK_SCIx(interface_id));
+		dummy = *MSTPCRA;
+	} else {
+		LOG_ERR("SCI modules in the secure domain on RZT2M are not supported.");
+		return -ENOTSUP;
+	}
+
+	/* Dummy-read at least five times as stated in 8.3.1 of the User's Manual: Hardware */
+	dummy = *RDR(config->base);
+	dummy = *RDR(config->base);
+	dummy = *RDR(config->base);
+	dummy = *RDR(config->base);
+	dummy = *RDR(config->base);
+
+	k_spin_unlock(&data->lock, key);
+	irq_unlock(irqkey);
+	return 0;
+}
+
+static int rzt2m_uart_init(const struct device *dev)
+{
+	if (!dev || !dev->config || !dev->data) {
+		return -ENODEV;
+	}
```

Comment:
This is not needed because the macro `UART_RZT2M_INIT()` below makes sure `dev` will not be NULL, as well as `dev->data` and `dev->config`.

Reply:
I've removed these lines. Thank you.